cmd/util/cmd/epoch-recovery/cmd/reset.go

```diff
@@ -0,0 +1,207 @@
+	}
+}
+
+// extractResetEpochArgs extracts the required transaction arguments for the `resetEpoch` transaction
+func extractResetEpochArgs(snapshot *inmem.Snapshot) cadence.Array {
+
+	// get current epoch
+	epoch := snapshot.Epochs().Current()
+
+	// set payout
+	payout := uint64(0)
+
+	// read random source from epoch
+	randomSource, err := epoch.RandomSource()
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not get random source from epoch")
+	}
+
+	// read first view
+	firstView, err := epoch.FirstView()
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not get first view from epoch")
+	}
+
+	// read final view
+	finalView, err := epoch.FinalView()
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not get final view from epoch")
+	}
+
+	// read collector clusters and convert to strings
+	clustering, err := epoch.Clustering()
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not get clustering from epoch")
+	}
+
+	// read dkg public keys for all participants
+	initialIdentities, err := epoch.InitialIdentities()
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not get initial identities from epoch")
+	}
+
+	dkg, err := epoch.DKG()
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not get dkg from epoch")
+	}
+
+	dkgKeys := make([]crypto.PublicKey, 0, len(initialIdentities))
+	for _, identity := range initialIdentities {
+		keyShare, err := dkg.KeyShare(identity.NodeID)
+		if err != nil {
+			log.Fatal().Str("node_id", identity.NodeID.String()).Err(err).Msg("coiuld get key share for node id")
+		}
+		dkgKeys = append(dkgKeys, keyShare)
+	}
+
+	// TODO: read in QCs
+
+	return convertResetEpochArgs(randomSource, payout, firstView, finalView, clustering, []string{}, dkgKeys)
+}
+
+// convertResetEpochArgs converts the arguments required by `resetEpoch` to cadence representations
+func convertResetEpochArgs(randomSource []byte,
+	payout, firstView, finalView uint64,
+	clustering flow.ClusterList, clusterQCs []string, dkgPubKeys []crypto.PublicKey) cadence.Array {
+
+	args := make([]cadence.Value, 0)
+
+	// add random source
+	args = append(args, cadence.NewString(hex.EncodeToString(randomSource)))
+
+	// add payout
+	cdcPayout, err := cadence.NewUFix64(fmt.Sprint(payout))
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not convert payout to cadence type")
+	}
+	args = append(args, cdcPayout)
+
+	// add first view
+	args = append(args, cadence.NewUInt64(firstView))
+
+	// add final view
+	args = append(args, cadence.NewUInt64(finalView))
+
+	// add clusters
+	cdcClusters := make([]cadence.Value, len(clustering))
+	for i, cluster := range clustering {
+		clusterNodeIDs := cluster.NodeIDs()
+		nodeIDStrings := make([]cadence.Value, len(clusterNodeIDs))
+		for j, nodeID := range clusterNodeIDs {
+			nodeIDStrings[j] = cadence.NewString(nodeID.String())
+		}
+		cdcClusters[i] = cadence.NewArray(nodeIDStrings)
+	}
+	args = append(args, cadence.NewArray(cdcClusters))
```

Comment:
The clusters need to be the `Cluster` type, not just arrays of strings. See the type definition in https://github.com/onflow/flow-core-contracts/blob/feature/epochs/contracts/epochs/FlowEpochClusterQC.cdc#L82-L121

Reply:
This might have caused some confusion because in the transaction I provided, I only used arrays of strings for the argument: https://github.com/onflow/flow-core-contracts/blob/feature/epochs/transactions/epoch/admin/reset_epoch.cdc#L9

Should I update that transaction to use the struct as an argument instead of the strings?

Reply:
Any reason why these are also empty? https://github.com/onflow/flow-core-contracts/blob/feature/epochs/transactions/epoch/admin/reset_epoch.cdc#L20-L22

Reply:
Sorry folks, I was looking at the function rather than the transaction.

Actually, the way the transaction is right now might save us some unnecessary work. The `collectorClusters`, `clusterQCs`, `dkgPubKeys` parameters are all ultimately unused. If we set these to empty lists within the transaction, and say that a reset epoch simply does not store this information, we save a good deal of complexity here. Let me know if you're OK with that @joshuahannan. If we do this, we can remove the `collectorClusters`, `clusterQCs`, `dkgPubKeys` parameters from the transaction.

Reply:
Another note: we don't currently check consistency of the epoch counter between the smart contract and the protocol state in this recovery flow. If the epoch counter is inconsistent, it will cause the chain to halt at the first epoch phase transition. 

Probably the easiest way to check this is to pass in the `epochCounter` to the `resetEpoch` method, and have the method revert if it does not match `FlowEpoch.currentEpochCounter` cc @joshuahannan

Reply:
Those both sound good to me. I'll create an issue